### PR TITLE
Create application start barrier

### PR DESCRIPTION
**Pull request by AndrewDJBanks**
See issue #12815

### CODE REVIEW

**Comment by AndrewDJBanks** [comment]
#build

**Comment by AndrewDJBanks** [comment]
#build

**Comment by AndrewDJBanks** [comment]
Note that: 
https://wasrtc.hursley.ibm.com:9443/jazz/resource/itemOid/com.ibm.team.build.BuildResult/_bpnpsLn4EeqsDL_UN9umug
contains an example of why the function is necessary. com.ibm.ws.jdbc.fat.db2.JDBCKerberosTest has run before  LTPA2 initialisation has completed.

File: dev/com.ibm.ws.app.manager/src/com/ibm/ws/app/manager/internal/lifecycle/ApplicationStartBarrierImpl.java [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+
+    @Reference(cardinality = ReferenceCardinality.MULTIPLE)
+    void setApplicationPrereq(ApplicationPrereq applicationPrereq) {
+    }
+
+    /**
+     * Non configurable immediate component to track prereqs as they appear.
+     */
+    @Component(configurationPolicy = ConfigurationPolicy.IGNORE, immediate = true)
+    public static class PrereqWatcher {
```
Comment (tjwatson) [inline dev/com.ibm.ws.app.manager/src/com/ibm/ws/app/manager/internal/lifecycle/ApplicationStartBarrierImpl.java:53]: It is not clear to me how this component does anything other than provide some additional trace points.

File: dev/com.ibm.ws.app.manager/src/com/ibm/ws/app/manager/internal/lifecycle/ApplicationStartBarrierImpl.java [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+            }
+
+            Arrays.sort(prereqClassNames);
+            return Arrays.toString(prereqClassNames);
+        }
+
+        @Reference(cardinality = ReferenceCardinality.MULTIPLE, policy = ReferencePolicy.DYNAMIC, policyOption = ReferencePolicyOption.GREEDY)
+        void setApplicationPrereq(ApplicationPrereq applicationPrereq) {
+            realisedPrereqs.add(applicationPrereq.getClass().getName());
+            Collections.sort(realisedPrereqs);
```
Comment (tjwatson) [inline dev/com.ibm.ws.app.manager/src/com/ibm/ws/app/manager/internal/lifecycle/ApplicationStartBarrierImpl.java:94]: Why always sort even when trace is not enabled.  I only see realisedPrereqs getting used within trace blocks.

File: dev/com.ibm.ws.app.manager.lifecycle/src/com/ibm/wsspi/application/lifecycle/ApplicationPrereq.java [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+ * Copyright (c) 2020 IBM Corporation and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *     IBM Corporation - initial API and implementation
+ *******************************************************************************/
+package com.ibm.wsspi.application.lifecycle;
```
Comment (tevans78) [inline dev/com.ibm.ws.app.manager.lifecycle/src/com/ibm/wsspi/application/lifecycle/ApplicationPrereq.java:11]: @tjwatson makes a good point, you can't expose this as a SPI straight away. You need a Feature and UFO to expose a new SPI. Please change the package name.

**Comment by tjwatson** [comment]
If we go with the approach to avoid calling application handlers until the ApplicationStartBarrier is available we could place the following reference on `com.ibm.ws.app.manager.internal.ApplicationConfigurator` and remove that reference from all Application handlers.

```
    @Reference
    private void setApplicationStartBarrier(ApplicationStartBarrier applicationStartBarrier) {
    }
```

We need to understand how this solution behaves when you dynamically enable a feature that adds a new `ApplicationPrereq`.  Does that take down `ApplicationStartBarrier` and do we gracefully shutdown existing applications?  Then successfully restart when the `ApplicationPrereq` is available again?

What if we unconfigure a feature providing a `ApplicationPrereq` and at the same time configure a different feature that provides another `ApplicationPrereq` such that the overall count of configured `ApplicationPrereq` remains the same?

**Comment by AndrewDJBanks** [comment]
#build

**Comment by AndrewDJBanks** [comment]
#build

**Comment by AndrewDJBanks** [comment]
#build

**Comment by AndrewDJBanks** [comment]
#build

File: dev/com.ibm.ws.app.manager/src/com/ibm/ws/app/manager/internal/lifecycle/ApplicationStartBarrierImpl.java (hunk 2 above)
Comment (chackoj) [reply dev/com.ibm.ws.app.manager/src/com/ibm/ws/app/manager/internal/lifecycle/ApplicationStartBarrierImpl.java:94]: Very good question. Would it be acceptable to use a sorted set instead, or should we simply defer sorting until we know it needs to be used?

**Comment by AndrewDJBanks** [comment]
#build

File: dev/com.ibm.ws.app.manager/src/com/ibm/ws/app/manager/internal/lifecycle/ApplicationStartBarrierImpl.java (hunk 2 above)
Comment (tjwatson) [reply dev/com.ibm.ws.app.manager/src/com/ibm/ws/app/manager/internal/lifecycle/ApplicationStartBarrierImpl.java:94]: I suppose a sorted set would be the most simple.  The number of prereqs likely is not high so sorting as you add likely should not show any measurable performance difference.

**Comment by AndrewDJBanks** [comment]
#build

**Comment by AndrewDJBanks** [comment]
#run-libby-bot

**Comment by AndrewDJBanks** [comment]
#build

File: dev/com.ibm.ws.app.manager/src/com/ibm/ws/app/manager/internal/lifecycle/ApplicationStartBarrierImpl.java (hunk 1 above)
Comment (AndrewDJBanks) [reply dev/com.ibm.ws.app.manager/src/com/ibm/ws/app/manager/internal/lifecycle/ApplicationStartBarrierImpl.java]: The test case and https://github.com/OpenLiberty/open-liberty/pull/11152
are (going to be)  applicationPrereqs

File: dev/com.ibm.ws.app.manager.lifecycle/src/com/ibm/wsspi/application/lifecycle/ApplicationPrereq.java (hunk 3 above)
Comment (AndrewDJBanks) [reply dev/com.ibm.ws.app.manager.lifecycle/src/com/ibm/wsspi/application/lifecycle/ApplicationPrereq.java:11]: we're not making it an SPI. (Yet)

**Review by brenthdaniel — APPROVED** [review]

**Pull request merged**